File: pset5.cpp
#include <iostream>
#include <random>
#include <cmath>

// Function to compute modular exponentiation (a^b mod n)
long long power(long long a, long long b, long long n) {
    long long result = 1;
    while (b > 0) {
        if (b % 2 == 1) {
            result = (result * a) % n;
        }
        a = (a * a) % n;
        b /= 2;
    }
    return result;
}

int main() {
    // Initialize values
    int u = 775198773;
    int t = 4;
    long long n = 12403180369;
    int power_of_two = std::pow(2, t);

    // Seed the random number generator
    std::random_device rd;
    std::mt19937 gen(rd());

    // Define the range for choosing a random base a
    int lower_bound = 1;
    int upper_bound = n;

    // Generate a random integer within the range
    std::uniform_int_distribution<> dis(lower_bound, upper_bound);
    int a = dis(gen);

    // Check for the property
    bool found = false;
    for (int i = 1; i <= t; ++i) {
        long long exp = power(power_of_two, i - 1, n);
        long long result = power(a, exp * u, n);
        if (result != 1 && result != n - 1) {
            found = true;
            std::cout << "Witness found: " << a << std::endl;
            break;
        }
    }

    // If no witness is found, it's highly likely that n is prime
    if (!found) {
        std::cout << "No witness found. " << n << " is likely prime." << std::endl;
    }

    return 0;
}

File: pset5.py
import random

# n value that we are testing 
n = 63973

# Calculate GCD
def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

# Finds u and t according to the notes:  n - 1 = 2^t * u
def find_u_and_t(n):
    u = n - 1
    t = 0
    while u % 2 == 0:
        u //= 2
        t += 1
    return u, t

# Find u and t
u, t = find_u_and_t(n)

# Used for Rabin's alg
def find_real_a(a, i):
    power = pow(2, i - 1) * u
    real_a = pow(a, power, n)
    print(real_a)
    return

# Fermat's test + Rabin's test
def notesAlg(n):
    for _ in range(100):
        a = random.randint(1, n)


        # Uses Fermat's Test
        if pow(a, n - 1, n) != 1:
            if gcd(a, n) == 1:
                print(f"n is composite by Fermat's. Witness: {a}")
                return None

        # If Fermat's fails, moves onto Rabin's test from notes
        for i in range(1, t + 1):
            cPower = pow(2, i)
            res1 = pow(a, cPower * u, n)
            if res1 == 1:
                newcPower = pow(2, i - 1)
                res2 = pow(a, newcPower * u, n)
                if res2 != 1 and res2 != -1:
                    if gcd(a, n) == 1:
                        print(f"n is composite by Rabin's. Witness: {a} with power {i}")
                        print(f'REAL A: {res2}')
                        find_real_a(a, i)
                        return None
        
notesAlg(n)







"""
def is_witness(a, n):
    r1 = pow(a, n-1, n) != 1
    r2 = pow(a, (n-1) // 2, n) != 1 and pow(a, (n-1) // 2, n) != n - 1
    print(r1)
    print(r2)
    return r1 or r2

found_witness = False

for i in range(t):
    # Calculate the power of 2 times u
    exp = pow(2, i) * u
    
    # Compute a^(2^{i}u) mod n
    result = pow(a, exp, n)
    
    # Check the conditions
    if result != 1 and result != n - 1:
        found_witness = True
        break

# Check if a witness was found
if found_witness:
    print(f"A witness was found for n = {n}: a = {a}")
else:
    print("No witness was found for n =", n)
    
is_witness(a, n)

def find_u_and_t(n):
    u = n - 1
    t = 0
    while u % 2 == 0:
        u //= 2
        t += 1
    print(u, t)
    return u, t

"""

    